---
title: "similaRpeak: similarity between two profiles"
output:
  html_document:
    theme: null
    toc: yes
  pdf_document:
    theme: null
    toc: yes
---

<!-- rmarkdown v1 -->
<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{similaRpeak: similarity between two profiles}
-->


Metrics which estimate similarity between two ChIP profiles
====================================================================

Astrid Louise Deschenes, Elsa Bernatchez, Charles Joly Beauparlant, 
Fabien Claude Lamaze, Rawane Samb, Pascal Belleau and Arnaud Droit.

This package and the underlying similaRpeak code are distributed 
under the Artistic license 2.0. You are free to use and redistribute 
this software. 


## Quick Start

Enhancer regions found in the mouse embryonic stem cells thought to be 
important for the cell fate as being highly active and connected regions have
been selected (Whyte et al. 2013 & Dowen et al. 2014). Using Encyclopedia of 
DNA Elements (ENCODE) data (Dunham I et al. 2012), the epigenetic landscape 
overlap surrounding those enhancer regions at two histone 
post-transcriptional modifications linked to highly active enhancers H3K27ac 
(DCC accession: ENCFF000ASG) and H3K4me1 (DCC accession: ENCFF000ARY) for human
specie has been studied. 

First, the `similaRpeak` package nust be loaded.

```{r libraryLoad}
suppressMessages(library(similaRpeak))
```

A region, chr7:61968807-61969730, shows interesting profiles for both histones.
Let's load the data for this region.

```{r profiles, collapse=T}
load(system.file("extdata/chr7_profiles.RData", package = "similaRpeak"))

head(H3K27ac_profile, n=15)
head(H3K4me1_profile, n=15)
```

H3K27ac and H3K4me1 profiles have those shapes:

```{r graphProfiles, echo=FALSE, fig.align='center', fig.height=6 }
plot(H3K27ac_profile, type="l", col="blue", xlab="", ylab="", ylim=c(0, 700))
par(new=TRUE)
plot(H3K4me1_profile, type="l", col="darkgreen", xlab="Position", 
     ylab="Coverage", ylim=c(0, 700))
legend("topleft", c("H3K27ac","H3K4me1"), cex=1.2, 
            col=c("blue","darkgreen"), lty=1)
```

The metrics are calculated using the `similarity` function which takes as 
arguments the two ChIP profiles vectors and the threshold values.

```{r metricCalculation}
metrics <- similarity(H3K27ac_profile, 
                            H3K4me1_profile, 
                            ratioAreaThreshold=5, 
                            ratioMaxMaxThreshold=2, 
                            ratioIntersectThreshold=5, 
                            ratioNormalizedIntersectThreshold=2,
                            diffPosMaxThresholdMinValue=10, 
                            diffPosMaxThresholdMaxDiff=100, 
                            diffPosMaxTolerance=0.01)
```

The `similarity` function returns a list which contains the general 
information about both ChIP-Seq profiles and a list of all four metrics. 

```{r metricReturn, collapse=T}
metrics
```

Each specific information can be directly accessed. Some examples:

```{r getInfo, collapse=T}
metrics$areaProfile1
metrics$areaProfile2
metrics$metrics$RATIO_INTERSECT
```

The **RATIO_INTERSECT** value of `r round(metrics$metrics$RATIO_INTERSECT, 2)` 
and the **RATIO_MAX_MAX** value of `r round(metrics$metrics$RATIO_MAX_MAX, 2)` 
are quite low. Both values can be explained by the large difference in 
coverage between profiles. Those values could be interpreted as two profiles 
with low level of similarity. However, the **RATIO_NORMALIZED_INTERSECT** of 
`r round(metrics$metrics$RATIO_NORMALIZED_INTERSECT, 2)` is much closer to 1. It 
could be a sign that the profiles, once normalized, are quite similar.
This hypothesis can be validated by looking at a graph of the normalized
profiles :

```{r graphProfilesNorm, echo=FALSE, fig.align='center', fig.height=6 }
plot(H3K27ac_profile*length(H3K27ac_profile)/sum(H3K27ac_profile, na.rm=T), 
     type="l", col="blue", xlab="", ylab="", ylim=c(0, 3.5))
par(new=TRUE)
plot(H3K4me1_profile*length(H3K4me1_profile)/sum(H3K4me1_profile, na.rm=T), 
     type="l", col="darkgreen", xlab="Position", 
     ylab="Normalized Coverage (Coverage/Mean Coverage)",  ylim=c(0, 3.5))
legend("topleft", c("H3K27ac","H3K4me1"), cex=1.2, 
            col=c("blue","darkgreen"), lty=1)
```


## Introduction

The **similaRpeak** package calculates metrics to estimate the level 
of similarity between two ChIP profiles.

The metrics are:

* **RATIO_AREA**: The ratio between the profile areas. 
The first profile is always divided by the second profile. `NA` is returned 
if minimal area threshold is not respected for at least one of the profiles.

* **DIFF_POS_MAX**: The difference between the maximal peaks positions. The 
difference is always the first profile value minus the second profile value. 
`NA` is returned if minimal peak value is not respected. A profile can have 
more than one position with the maximum value. In that case, the median 
position is used. A threshold argument can be set to consider all positions 
within a certain range of the maximum value. A threshold argument can also be 
set to ensure that the distance between two maximum values is not too wide. 
When this distance is not respected, it is assumed that more than one peak 
is present in the profile and `NA` is returned. 

* **RATIO_MAX_MAX**: The ratio between the peaks values 
in each profile. The first profile is always divided by the second profile. `NA` 
if minimal peak value threshold is not respected for at least one of the 
profiles.

* **RATIO_INTERSECT**: The ratio between the intersection area and the total 
area of the two profiles. `NA` if minimal area threshold is not respected for 
the intersection area.
   
* **RATIO_NORMALIZED_INTERSECT**: The ratio between the intersection area and 
the total area of two normalized profiles. The profiles are normalized by 
divinding them by their average value. `NA` if minimal area threshold is not 
respected for the intersection area.
 
## Inputs

### ChIP profiles vectors

To estimate the level of similarity between two ChIP profiles for a specific 
region, two `vector` containing the RPM values of the two ChIP profiles for 
each position of the selected region. Both `vector` should have the same length 
and should not contain any negative value.

### Threshold variables

When the metric includes a ratio calculation, a threshold for the denominator 
is assigned to the metric. The threshold has to be positive numerical value. 
When the threshold is not respected, the metric is not calculated and the 
`NA` value is assigned to it.

The threshold variables are:

* **ratioAreaThreshold**: The minimum denominator accepted to calculate 
the ratio of the area between both profiles. Default = 1.

* **ratioMaxMaxThreshold**: The minimum denominator accepted to calculate the 
ratio of the maximal peaks values between both profiles. Default = 1.

* **ratioIntersectThreshold**: The minimum denominator accepted to calculate 
the ratio of the intersection area of both profiles over the total area. 
Default = 1.

* **ratioNormalizedIntersectThreshold**: The minimum denominator accepted to 
calculate the ratio of the intersection area of both profiles over the total 
area for normalized profiles. Default = 1.

* **diffPosMaxThresholdMinValue**: The smaller maximum value accepted to 
calculate the metric. When this value is not respected, it is assumed that no 
peak is present in the profile. Default = 1.

* **diffPosMaxThresholdMaxDiff**: The larger distance accepted between 2 
maximum positions in one profile to calculate the metric. When this distance 
is not respected, it is assumed that more than one peak is present in the 
profile. Default=100.

* **diffPosMaxTolerance**: The maximum variation accepted on the maximum value 
to include a position in the maximum positions list and use it to calculate the
metric. Default=0.01.


## Metric definition

Mathematically, a metric is considered as a function that quantifies the 
similarity between two objects.The function must return zero when the two 
objects are perfectly similar (identity of indiscernibles) and a non-negative 
value when are dissimilar. 

The metrics present in the **similaRpeak** package do not strictly respect this
standard but can all be translated to pseudometrics. A pseudometric is a 
function d which satisfies the axioms for a metric, except that instead of the
identity of indiscernibles axiom, the metric must only return zero when it 
compare an object to itself.

By using the absolute value of the **DIFF_POS_MAX** metric, the definition of 
a pseudometric is formally respected. However, the respective position of the 
maximum peak of profiles is lost.

$$ |DIFF\_POS\_MAX| $$

By using the absolute value of the logarithm of the 
**RATIO_AREA**, **RATIO_MAX_MAX**, **RATIO_INTERSECT** and 
**RATIO_NORMALIZED_INTERSECT** metrics, the definition of a pseudometric is 
formally respected.

$$ \log(RATIO) $$


## RATIO_AREA Metric






## Metrics calculation using MetricFactory object
 
It is possible to create only one selected metric by using the 
**MetricFactory** object (with the possibility of specifying the thresholds) 
and by passing the name of the metric to create 
(**RATIO_AREA**, **DIFF_POS_MAX**, **RATIO_MAX_MAX**, **RATIO_INTERSECT** or 
**RATIO_NORMALIZED_INTERSECT**):

```{r factory}
factory = MetricFactory$new(diffPosMaxTolerance=0.04)
```

The factory has to be iniatized only once and can be used as many times as 
necessary. It can calculate the same metrics but with different profiles or 
different metrics with same profiles as long as the thresholds values 
remain the same:

```{r factoryDemo, collapse=T }
ratio_max_max <- factory$createMetric(metricType="RATIO_MAX_MAX", 
                                        profile1=c(1,59,6,24,65,34,15,4,53,22), 
                                        profile2=c(15,9,46,44,9,39,27,34,34,4))

ratio_max_max

ratio_normalized_intersect <- factory$createMetric(metricType="RATIO_NORMALIZED_INTERSECT", 
                                        profile1=c(1,59,6,24,65,34,15,4,53,22), 
                                        profile2=c(15,9,46,44,9,39,27,34,34,4))
ratio_normalized_intersect
```


## References

Dowen JM, Peng Fan Z, Hnisz D, Ren G, et al. (2014) Control of Cell Identity 
Genes Occurs in Insulated Neighborhoods in Mammalian Chromosomes.
Cell 159:374-387. doi: 10.1016/j.cell.2014.09.030

Dunham I, Kundaje A, Aldred SF, et al. An integrated encyclopedia of DNA 
elements in the human genome. Nature. 2012 Sep 6;489(7414):57–74.

Whyte WA, Orlando DA, Hnisz D, Abraham BJ, Lin CY, et al. (2013) Master 
transcription factors and mediator establish super-enhancers at key cell 
identity genes. Cell 153:307–319. doi: 10.1016/j.cell.2013.03.035
